Spw grid view2

will-moore opened the pull request:
# Spatial view of fields in Wells

2nd PR working towards https://github.com/openmicroscopy/design/issues/52
Builds on https://github.com/openmicroscopy/openmicroscopy/pull/4863

Also see https://trello.com/c/GxqDnD0P/211-spw-follow-up

To test:
 - Browse SPW, click on Well. Well 'bird's eye' spatial view should open below tree.
 - Selecting images in central lower panel should highlight these in bird's eye view.
 - Selecting images in bird's eye by clicking / shift-clicking or dragging should highlight same images in central lower panel and right panel.
 - Bird's eye panel will stay open until X in top right corner is clicked.
 - Will be shown again when Well is clicked.
 - Enter a value in the "Wrap [ ] images per row" field. On 'Enter' this should wrap the images in each well below. Should also wrap any new wells loaded.
 - Remove wrapping by clearing the input (or enter 0).

TODO:
- [x] integration tests for posX and posY 

![screen shot 2016-11-14 at 16 46 02](https://cloud.githubusercontent.com/assets/900055/20295500/361a85f8-aafd-11e6-9d4a-8c5f219c19ed.png)


![screen shot 2016-11-14 at 16 19 49](https://cloud.githubusercontent.com/assets/900055/20272589/48fcb96a-aa86-11e6-8633-f1db0ae178fe.png)

will-moore committed:
Add 'position' data to WellImages_json

will-moore committed:
Move scripts from plate.html to base_container - remove duplicates

will-moore committed:
Move functions from plate.html to spwgridview.js

will-moore committed:
Basic display of images in well bird's eye view

will-moore committed:
Sync selection of images in well birds eye

will-moore committed:
Improving Well birds eye view

will-moore committed:
More fixes and tweaks to well_birds_eye

will-moore committed:
Hide well_birds_eye on click X

jburel commented:
gh-4863 has been merged
so better to add your commits on top of develop so we can follow what's going on

will-moore commented:
Data generated with script from https://github.com/dsudar/OMEROscripts/pull/1

![screen shot 2016-11-04 at 11 09 31](https://cloud.githubusercontent.com/assets/900055/20003603/54460998-a27f-11e6-9d9c-2e9cf64a4bc8.png)

will-moore committed:
Fix height of center panel .tableLayout

will-moore committed:
Move styles from plate.html to layout.css

will-moore committed:
Adding 'Wrap x images per row' functionality

will-moore committed:
Robot setup populates posX and posY on WellSamples

will-moore committed:
Fix PYTHONPATH for wellSample posXY script in robot_setup.sh

will-moore committed:
Robot tests grid layout of well-images

will-moore committed:
robot test typo fix

gusferguson commented:
@will-moore 

Tested using http://10.0.51.106/web/webclient/userdata/?experimenter=2 user-1 jpeg-xr group Safari

2007.10.17.17.04.55.xdce
Plate ID: 2
Owner: user 1

and other data.

Birdseye view shows spatial arrangement of fields when selected.

In the birdseye shift-click for multi selection does not work in Safari/Firefox/Chrome but cmd-click and click-drag both work.

All highlighting appears appropriate.
Small point - highlighting around image/field thumbnails in bottom half of centre pane is not really prominent enough - difficult to see which thumbnails are highlighted - particularly when non-adjacent multiple selections are made via birdseye. Could it be made darker or thickness increased?
Birdsye closes and reappears as expected.

Wrapping of thumbnails in well rows behaves as expected.
Did have single episode where clicked into wrap text box and incremented using ^ arrow on right side - cursor became trapped in wrap text box -and could select images and wells in centre pane, but could not select images in birdseye. Freed up on clicking to another well and could not replicate subsequently so must some issues between wrapping text box and birdseye view - but cannot be more helpful in pinning it down..

Could we reduce or remove the corner rounding on the wrap text box? - because it is so short it looks oval

All other functionality works as expected.

![webclient](https://cloud.githubusercontent.com/assets/1552703/20096857/9fd83e40-a5a3-11e6-8c3f-aeea526bf91c.png)

Good to merge from my POV.

atarkowska commented:
Robots are failing in https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/438/console. Where does `wsScript.py` come from exactly?

will-moore committed:
robot_setup passes 'host' and 'port' to BlitzGateway in wsScript.py

will-moore committed:
Increase select highlight width around well images

will-moore committed:
Handle ?show=image-1 to select well and image

will-moore committed:
Add test of listwellimages json with posX & posY

will-moore committed:
Adding to robot 'Test Spw Grid Layout' test

will-moore committed:
Fix ?show=well-1 showing Well in right panel

will-moore committed:
flake8 fix

will-moore committed:
Robot test for 'Show Well Images'

will-moore committed:
Added new robot 'Test Spw Spacial Birds Eye'

will-moore commented:
- Fixed highlighted border width for thumbnails in bottom panel.
- Added robot tests & integration test
- Added support for webclient/?show=image-1|image-2.  To test, select a Well, then select several images in the bottom panel. Copy the "url" from the link button in right panel, paste into browser - page loaded should return to selecting the same images.
- Same should work for selecting multiple Wells.

gusferguson commented:
@will-moore 

Tested using http://web-dev-merge.openmicroscopy.org/webclient/ user-1, user-3, user-5

URL from link when multiple wells or images are selected works as expected and takes you to the selected objects.
More prominent highlight border for thumbnails looks good.

Noticed that some data e.g. http://web-dev-merge.openmicroscopy.org/webclient/?show=well-240 has four fields and they obviously don't have spatial metadata so they all stack up in the centre of the LHP spatial birdseye - which is expected. Yet other plates e.g. http://web-dev-merge.openmicroscopy.org/webclient/?show=well-41 have multiple fields yet nothing shows at all in theLHP spatial birdseye. 
Is this just an artefact of the constructed plates we are testing with or is this an issue?

Good to merge from my POV.

will-moore committed:
Filter listWellImages by 'run'

will-moore committed:
Pass acquisition/run to listWellImages url

will-moore committed:
Robot tests plate has 2 acquisitions

And also a smaller plate where we don't need multiple acquisitions

will-moore commented:
The last commits add a fix and test for recent bug. To test:

Import SPW with multiple acquisitions.
E.g.

```
$ touch "test&plates=1&plateAcqs=2&plateRows=2&plateCols=3&fields=5&screens=0.fake"
```

Browse to plate, select well and check that only well Images from the current Acquisition are shown in the bottom panel.
Compare with Insight - are the same images loaded in bottom panel?

jburel commented on components/tests/ui/robot_setup.sh:
```diff
@@ -45,6 +47,34 @@ echo "A2,Treatment,10" >> "$BULK_ANNOTATION_CSV"
 # Create file for upload as File Annotation
 echo "Robot test file annotations" > "$FILE_ANNOTATION"
 
```
could you close the gateway when you are done with wsScript.py?

will-moore replied:
If I call ```conn.close()``` at the end of wsScript.py then this kills the session that I have joined and the rest of the setup script fails.
The wsScript.py doesn't create any stateful services and the whole session is completed at the end of the setup script with ```$ bin/omero logout``` so I don't think we need to close the BlitzGateway in wsScript.py.

jburel replied:
so it is similar to the problem I reported here
https://trello.com/c/yc8YNhUa/117-omero-web-kills-running-script-on-logout
I think my comment was added before I discovered that issue

jburel commented on components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/data/containers.html:
```diff
@@ -845,6 +845,14 @@
 
     </div>
+
+    <!-- Panel hidden unless needed for showing spacial birds eye view of Wells - see ome.spwgridview.js -->
```
spatial

jburel commented on components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/data/includes/center_plugin.thumbs.js.html:
```diff
@@ -811,6 +812,10 @@
         $contentDetails.removeData('path');
         $("div#page").removeData('page');
         $contentDetails.removeData('field');
+        // empty SPW 'spacial' birds eye view below tree
```
same

jburel commented on components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/data/plate.html:
```diff
@@ -249,71 +197,83 @@
 
+            // Lower panel.....
+
+            // Wrap images to fixed number per row...
```
fix

jburel commented on components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/data/plate.html:
```diff
@@ -249,71 +197,83 @@
             });
 
-            var showImagesInWell = function() {
+            // Callback for the spacial birds eye component
```
same

jburel commented on components/tools/OmeroWeb/test/integration/test_plategrid.py:
```diff
@@ -67,10 +69,14 @@ def update_service(client):
 def well_sample_factory(itest):
 
-    def make_well_sample():
+    def make_well_sample(xPos=None, yPos=None):
```
x_pos, y_pos

jburel commented:
Could you add tests for methods modified in ``view.py``? e.g. ``load_plate``

will-moore committed:
Fix xPos, yPos -> x_pos, y_pos in test_plategrid.py

will-moore commented:
@jburel - the minor change in ```def load_plate()``` results in a small change in the html/js of the plate page which is needed to filter wellImages by 'acquisition'. That functionality is tested by robot tests (now that the plate we're creating in robot tests has 2 acquisitions. https://github.com/openmicroscopy/openmicroscopy/pull/4924/commits/80377b640d44b5891b458c91b79e4e7ec91febdc

will-moore committed:
Typo fixes

will-moore commented:
Currently robot test is failing to select Well for ```webclient/?show=well-8944|well-8946``` when those Wells are in multiple Acquisitions:
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/459/robot/Web%20&%20Web/Web_1/Show%20Test/Test%20Show%20Plate/

Turns out that there are lots of parts to this bug:
In ```show.py``` we handle multiple objects differently from single object
```
        # support for multiple objects selected by ID,
        # E.g. show=image-1|image-2
        if 'id' in attributes.keys() and len(self._initially_select) > 1:
            # 'image.id-1' -> 'image-1'
            self._initially_select = [
                i.replace(".id", "") for i in self._initially_select]
        else:
            # Only select a single object
            self._initially_select = self._initially_open[:]
```
So, the initially_select is ```[u'well-8944', u'well-8946']``` instead of ```['acquisition-4594', u'well-8944']```. 
Then the JavaScript code tries to get ```path_to_object/?well=8944``` instead of ```?acquisition-4594```.
This returns many paths to Well, since we get a path for every ```WellSample``` for that Well (E.g. 2 'runs' and '5 fields' each is 10 paths.
In ```ome.tree.js``` a loop tries to traverse ALL of these paths and select the 'acquisition` for each. If 2 or more are selected in the tree, the centre panel is empty (hence no Well selected).

will-moore committed:
Don't try to 'show' multiple Wells in show._initially_select

See https://github.com/openmicroscopy/openmicroscopy/pull/4924#issuecomment-263878218

will-moore committed:
Revert "Don't try to 'show' multiple Wells in show._initially_select"

This reverts commit 05915ac6a401aee87253a0acd2f85e7b7432a5cb.

will-moore committed:
Use both show.initially_open and show.initially_select

to browse the jsTree and select target objects respectively.
See https://github.com/openmicroscopy/openmicroscopy/pull/4924#issuecomment-263878218
When we have show=well-1|well-2 we need to initially_open a single acquisition but
also initially_select both wells.

will-moore committed:
Fix robot test

will-moore committed:
When ?show=well-1 only show Well when plate first loaded

will-moore committed:
Merge remote-tracking branch 'origin/develop' into spw_grid_view2

will-moore commented:
To test, check robot test is passing (see failing test link above https://github.com/openmicroscopy/openmicroscopy/pull/4924#issuecomment-263878218).
If you want to test manuall:
 - find a Plate with 2 or more acquisitions
 - Select an acquisition and select 2 or more wells
 - Click "Link icon" to get url webclient/?show=well-1|well-2 
 - Paste into new tab, check that you are navigated to the same place and both well are selected.
 - If you load a different Plate, then go back to the original one, it **shouldn't** select the wells when it loads a second time. cc @chris-allan @aleksandra-tarkowska

will-moore commented:
Robot test passing at https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/464/robot/Web%20&%20Web/Web_1/Show%20Test/Test%20Show%20Plate/

gusferguson commented:
@will-moore 

Tested using http://web-dev-merge.openmicroscopy.org/webclient/ user-5 and user-1

Behaves as expected.
No issues apparent.

Good to merge.

snoopycrimecop commented:
Conflicting PR. Removed from build [OMERO-DEV-merge-push#532](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/532/). See the [console output](https://ci.openmicroscopy.org/job/OMERO-DEV-merge-push/532/consoleText) for more details.
Possible conflicts:
  - PR #4694 will-moore 'filter rating'
    - components/tests/ui/resources/web/tree.txt

jburel commented on components/tests/ui/robot_setup.sh:
```diff
@@ -45,6 +47,34 @@ echo "A2,Treatment,10" >> "$BULK_ANNOTATION_CSV"
 
+# Create a python script for setting posX and posY on wellsamples
+# Used below after importing a plate
+WELLSCRIPT="wsScript.py"
```
It will be better to have the script in a separate file so it is easier to understand and potentially re-use if need be

jburel commented on components/tools/OmeroWeb/omeroweb/webclient/static/webclient/css/layout.css:
```diff
@@ -2198,6 +2200,67 @@ div.paging input.button_pagination {
+#wellImages li {
+    display: inline-block;
+}
+#wellImages ul,  #wellImages h1{
```
space

jburel commented:
few minor comments
`spacial` to change in few places

will-moore committed:
Moving robot_setup.sh wsScript.py -> well_sample_posXY.py

will-moore committed:
Add Usage docstring to well_sample_posXY.py

will-moore committed:
Typo fixes

will-moore commented:
Check typos and check robot tests (spw_test).

jburel commented on components/tests/ui/testcases/web/spw_test.txt:
```diff
@@ -71,7 +71,7 @@ Test Spw Grid Layout
 
-Test Spw Spacial Birds Eye
+Test Spw Spatial Birds Eye
     [Documentation]     Tests layout and clicking of 'spacial bird's eye' panel
```
one more ``spacial`` ``'spacial bird's``

jburel commented:
one more typo.
Thanks for extracting the script
the robot tests were not run today due to server being down
Check the output of the test before merging due to the change in testing set-up

If we have more scripts required for testing, we may want to create a ``scripts`` repo under resources

jburel committed:
Change path to script.

Commit:
Merge pull request #55 from jburel/spw_grid_view2

path to script

jburel committed:
Remove extra ")"

Commit:
Merge pull request #56 from jburel/spw_grid_view2

Remove extra ")"

jburel commented:
The commits above fix the setup issue. Other failures in the robot tests.

will-moore commented:
[Show test](https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/489/robot/Web%20&%20Web/Web_1/Show%20Test/) and [SPW test](https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/489/robot/Web%20&%20Web/Web_1/Spw%20Test/) are now passing.
https://ci.openmicroscopy.org/view/Failing/job/OMERO-DEV-merge-robotframework/489/robot/

atarkowska commented on components/tools/OmeroWeb/omeroweb/webclient/templates/webclient/base/base_container.html:
```diff
@@ -179,6 +183,7 @@
         // jsTree code in ome.tree.js and center panel code in center_plugin.thumbs.js.html uses initially_select
         // instead of browser URL since URL may be /webclient/?path=plate.name-barcode|well.name-A1
         WEBCLIENT.initially_select = [{% for o in initially_select %}"{{ o }}"{% if not forloop.last %},{% endif %}{% endfor %}];
+        WEBCLIENT.initially_open = [{% for o in initially_open %}"{{ o }}"{% if not forloop.last %},{% endif %}{% endfor %}];
```
As discussed recently, constructing array this way is inconvenient, see https://github.com/openmicroscopy/openmicroscopy/pull/5018

aleksandra-tarkowska commented:
@will-moore could you please cherry-pick https://github.com/openmicroscopy/openmicroscopy/pull/5018/ here? It is the only way I managed to pass non digits.

atarkowska commented:
@will-moore is this PR ready to be merged? Either could you please cherry-pick https://github.com/openmicroscopy/openmicroscopy/pull/5018 here or let's merge and I will fix the above in my PR. It is the only way I managed to pass non digits.

will-moore commented:
I think this is good to merge now, with the robot tests passing.
I'd prefer to review #5018 separately instead of making this PR even more complex.
cc @jburel
